tokio-util/src/sync/cancellation_token.rs

```diff
@@ -319,560 +190,323 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
+    /// Disconnects the given parent from all of its children.
+    ///
+    /// Takes a reference to [Inner] to make sure the parent is already locked.
+    ///
+    /// Returns the (now parentless) children.
+    fn disconnect_children(node: &mut Inner) -> Vec<Arc<TreeNode>> {
+        for child in &node.children {
+            let mut child = child.inner.lock().unwrap();
+            child.parent_idx = 0;
+            child.parent = None;
         }
-    }
 
-    /// Returns a snapshot of the current atomic state of the token
-    fn snapshot(&self) -> StateSnapshot {
-        StateSnapshot::unpack(self.state.load(Ordering::SeqCst))
+        std::mem::take(&mut node.children)
     }
 
-    fn atomic_update_state<F>(&self, mut current_state: StateSnapshot, func: F) -> StateSnapshot
+    /// Figures out the parent of the node and locks the node and its parent atomically.
+    ///
+    /// The basic principle of preventing deadlocks in the tree is
+    /// that we always lock the parent first, and then the child.
+    ///
+    /// Sadly, it's impossible to figure out the parent of a node without
+    /// locking it. To then achieve locking order consistency, the node
+    /// has to be unlocked before the parent gets locked.
+    /// This leaves a small window where we already assume that we know the parent,
+    /// but neither the parent nor the node is locked. Therefore, the parent could change.
+    ///
+    /// To prevent that this problem leaks into the rest of the code, it is abstracted
+    /// in this function.
+    ///
+    /// The locked child and optionally its locked parent, if a parent exists, get passed
+    /// to the `func` argument via (node, None) or (node, Some(parent)).
+    fn with_locked_node_and_parent<F, Ret>(node: &Arc<TreeNode>, func: F) -> Ret
     where
-        F: Fn(StateSnapshot) -> StateSnapshot,
+        F: FnOnce(MutexGuard<'_, Inner>, Option<MutexGuard<'_, Inner>>) -> Ret,
     {
-        let mut current_packed_state = current_state.pack();
+        let mut potential_parent = {
+            let locked_node = node.inner.lock().unwrap();
+            match locked_node.parent.clone() {
+                Some(parent) => parent,
+                // If we locked the node and its parent is `None`, we are in a valid state
+                // and can return.
+                None => return func(locked_node, None),
+            }
+        };
+
         loop {
-            let next_state = func(current_state);
-            match self.state.compare_exchange(
-                current_packed_state,
-                next_state.pack(),
-                Ordering::SeqCst,
-                Ordering::SeqCst,
-            ) {
-                Ok(_) => {
-                    return next_state;
-                }
-                Err(actual) => {
-                    current_packed_state = actual;
-                    current_state = StateSnapshot::unpack(actual);
+            potential_parent = {
```

Comment:
I don't think this scope is useful. Just put
```Rust
potential_parent = actual_parent;
```
at the end.

Reply:
Fixed